src/main/java/org/terasology/additionalRails/action/RailSwitchAction.java

```diff
@@ -1,74 +1,167 @@
     private Block railSwitchSignalOff;
     private Block railSwitchSignalOn;
+    private Block railSwitchCombinedOff;
+    private Block railSwitchCombinedOn;
+    private RailsUpdateFamily railFamily;
+    private RailsUpdateFamily invertFamily;
+
+    private enum ActionSource {ACTIVATE, SIGNAL};
 
     @Override
     public void initialise() {
         final BlockManager blockManager = CoreRegistry.get(BlockManager.class);
+        railSwitchLeverOff = blockManager.getBlock("AdditionalRails:railSwitchLeverOff");
+        railSwitchLeverOn = blockManager.getBlock("AdditionalRails:railSwitchLeverOn");
         railSwitchSignalOff = blockManager.getBlock("AdditionalRails:railSwitchSignalOff");
         railSwitchSignalOn = blockManager.getBlock("AdditionalRails:railSwitchSignalOn");
+        railSwitchCombinedOff = blockManager.getBlock("AdditionalRails:railSwitchCombinedOff");
+        railSwitchCombinedOn = blockManager.getBlock("AdditionalRails:railSwitchCombinedOn");
+        railFamily = (RailsUpdateFamily)blockManager.getBlockFamily("Rails:rails");
+        invertFamily = (RailsUpdateFamily)blockManager.getBlockFamily("Rails:railsTBlockInverted");
     }
 
-    @ReceiveEvent(components = {SignalConsumerStatusComponent.class, RailSwitchComponent.class})
-    public void switchAction(OnChangedComponent event, EntityRef entity) {
-        if (entity.hasComponent(BlockComponent.class)) {
-            SignalConsumerStatusComponent consumerStatusComponent = entity.getComponent(SignalConsumerStatusComponent.class);
-            Vector3i blockLocation = new Vector3i(entity.getComponent(BlockComponent.class).getPosition());
-            Block block = worldProvider.getBlock(blockLocation);
-
-            SignalConsumerComponent signalConsumerComponent = entity.getComponent(SignalConsumerComponent.class);
-            RailSwitchComponent railSwitchComponent = entity.getComponent(RailSwitchComponent.class);
-
-            if (block == railSwitchSignalOff && consumerStatusComponent.hasSignal) {
-                railSwitchSignalOn.setKeepActive(true);
-                worldProvider.setBlock(blockLocation, railSwitchSignalOn);
-                blockEntityRegistry.getBlockEntityAt(blockLocation).addOrSaveComponent(signalConsumerComponent);
-                if (!blockEntityRegistry.getBlockEntityAt(blockLocation).hasComponent(RailSwitchComponent.class)) {
-                    blockEntityRegistry.getBlockEntityAt(blockLocation).addOrSaveComponent(railSwitchComponent);
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchComponent.class})
+    public void activateAction(ActivateEvent event, EntityRef entity) {
+        RailSwitchComponent railSwitchComponent = entity.getComponent(RailSwitchComponent.class);
+
+        updateSwitch(event, entity, railSwitchComponent, !railSwitchComponent.isOn, ActionSource.ACTIVATE);
+    }
+
+    @ReceiveEvent(components = {BlockComponent.class, RailSwitchComponent.class, SignalConsumerStatusComponent.class})
```

Comment:
When introducing new methods (and classes, etc) if at all possible please try to add some javadoc. Bonus points for also adding it to existing undocumented stuff you're interacting with anyway, if you understand it well enough to doc :-)